File: script/compile.scala
#!/bin/sh
exec scala "$0" "$@"
!#

import java.io.File
import scala.reflect.io.Directory
import scala.collection.mutable.ArrayBuffer
import sys.process._

object Compile {
  def main(args: Array[String]): Unit = {
    // project
    val PRO : String = "cgraph"

    // directory tree
    val DIR : String = "."
    val INC : String = DIR + File.separator + "include"
    val SRC : String = DIR + File.separator + "src"
    val TST : String = DIR + File.separator + "test"
    val LIB : String = DIR + File.separator + "lib"

    val CC : String = "cc"
    var CFLAGS : String = "-std=c89 -Wall -pedantic -fPIC"
    val CSFLAGS : String = "-shared"

    val MODE : String = "debug"
    if (MODE.equals("debug")) {
      CFLAGS += " -g -DDEBUG"
    } else if (MODE.equals("release")) {
      CFLAGS += " -static -O2"
    }

    val AR : String = "ar"
    val ARFLAGS : String = "-rcs"



    val CFILES : Array[File] = (new File(SRC)).listFiles().filter(_.isFile).filter(_.toString.endsWith(".c"))

    val SCRIPT_NAME = (new File(getClass()).protectionDomain.codeSource.location.path).name
    if (args.length == 0) {
      if (!obj.exists()) {
        (new File(LIB)).mkdir()
      }
      val OFILES = ArrayBuffer[String]()
      for(file <- CFILES) {
        val patten = "\.c$"r
        val obj : String = pattern.replaceAll(file, ".o")
        println("compile ${file} to ${obj}")
        "${CC} ${CFLAGS} -I${INC} -c ${file} -o ${obj}" !
        OFILES += obj
      }
    } else if (args(0) == "test") {

    } else if (args(0) == "clean") {
      for(file <- CFILES) {
        val patten = "\.c$"r
        val obj : String = pattern.replaceAll(file ".o")
        println("clean ${obj}")
        if (obj.exists()) {
          obj.delete()
        }
      }
    } else if (args(0) == "distclean") {
      for(file <- CFILES) {
        val patten = "\.c$"r
        val obj : String = pattern.replaceAll(file, ".o")
        println("clean ${obj}")
        if (obj.exists()) {
          obj.delete()
        }
      }

      println("clean ${LIB}")
      if (LIB.exists()) {
        LIB.delete()
      }
    } else if (args(0) == "help") {
      println("${SCRIPT_NAME} <target>")
      println("<target>: ")
      println("                    compile cgraph")
      println("          test      test cgraph\n")
      println("          clean     clean all the generated files")
      println("          distclean clean all the generated files and directories")
      println("          help      commands to this PROgram")
    } else {
      println("${args(0)} is an unsupported command")
      println("use \"${SCRIPT_NAME} help\" to know all supported commands")
    }
  }
}

File: script/compile.r
#!/usr/bin/Rscript

PRO <- "cgraph"
DIR <- "."
INC <- file.path(DIR, "include", fsep=.Platform$file.sep)
SRC <- file.path(DIR, "src", fsep=.Platform$file.sep)
TST <- file.path(DIR, "test", fsep=.Platform$file.sep)
LIB <- file.path(DIR, "lib", fsep=.Platform$file.sep)

CC <- "cc"
CFLAGS <- "-pedantic -Wall -fPIC -std=c89"
CSFLAGS <- "-shared"

MODE <- "debug"
if(MODE == "debug") {
  CFLAGS <- paste(CFLAGS, "-g -DDEBUG", sep=" ")
} else if(MODE == "release") {
  CFLAGS <- paste(CFLAGS, "-static -O2", sep=" ")
}

# package shared library
AR <- "ar"
ARFLAGS <- "-rcs"

CFILES <- lapply(list.files(SRC, pattern="\\.c$"), function(x) { file.path(SRC, x, fsep=.Platform$file.sep)})

if(.Platform$OS.type == "windows") {
  # target files
  LIBSHARED <- file.path(LIB, paste("lib", PRO, ".dll", sep=""), fsep=.Platform$file.sep)
  LIBSTATIC <- file.path(LIB, paste("lib", PRO, ".a", sep=""), fsep=.Platform$file.sep)
  # test files
  TSTFILE <- file.path(TST, paste(PRO, ".c", sep=""), fsep=.Platform$file.sep)
  TSTTARGET <- file.path(TST, paste(PRO, ".exe", sep=""), fsep=.Platform$file.sep)
} else {
  # target files
  LIBSHARED <- file.path(LIB, paste("lib", PRO, ".so", sep=""), fsep=.Platform$file.sep)
  LIBSTATIC <- file.path(LIB, paste("lib", PRO, ".a", sep=""), fsep=.Platform$file.sep)
  # test files
  TSTFILE <- file.path(TST, paste(PRO, ".c", sep=""), fsep=.Platform$file.sep)
  TSTTARGET <- file.path(TST, PRO, fsep=.Platform$file.sep)
}

args <- commandArgs(trailingOnly = TRUE)
script_name <- unlist(strsplit(commandArgs()[4], split="="))[2]
if(length(args) == 0) {
  if(!file.exists(LIB)) {
    dir.create(LIB)
  }
  OFILES <- character()
  for(file in CFILES) {
    obj <- gsub("\\.c$", ".o", file)
    print(sprintf("compile %s to %s", file, obj))
    system(sprintf("%s %s -I%s -c %s -o %s", CC, CFLAGS, INC, file, obj))
    OFILES <- append(OFILES, obj, after=length(OFILES))
  }
  print(sprintf("compile %s", LIBSHARED))
  system(sprintf("%s %s -o %s %s", CC, CSFLAGS, LIBSHARED, paste(OFILES, collapse=" ")))
  print(sprintf("compile %s", LIBSTATIC))
  system(sprintf("%s %s %s %s", AR, ARFLAGS, LIBSTATIC, paste(OFILES, collapse=" ")))
} else if (args[1] == "test") {
  print(sprintf("compile %s to %s", TSTFILE, TSTTARGET))
  system(sprintf("%s %s -I%s -o %s %s -L%s -static -l%s -lm", CC, CFLAGS, INC, TSTTARGET, TSTFILE, LIB, PRO))
  print(sprintf("test %s with %s", TSTTARGET, file.path(TST, "elements.csv", fsep=.Platform$file.sep)))
  system(sprintf("%s %s", TSTTARGET, file.path(TST, "elements.csv", fsep=.Platform$file.sep)))
} else if (args[1] == "clean") {
  for(file in CFILES) {
    obj <- gsub("\\.c$", ".o", file)
    print(sprintf("clean %s", obj))
    unlink(file.path(SRC, obj, fsep=.Platform$file.sep), force=TRUE)
  }
  print(sprintf("clean %s", LIBSTATIC))
  unlink(LIBSTATIC, force=TRUE)
  print(sprintf("clean %s", LIBSHARED))
  unlink(LIBSHARED, force=TRUE)
  print(sprintf("clean %s", TSTTARGET))
  unlink(TSTTARGET, force=TRUE)
} else if (args[1] == "distclean") {
  for(file in CFILES) {
    obj <- gsub("\\.c$", ".o", file)
    print(sprintf("clean %s", obj))
    unlink(file.path(SRC, obj, fsep=.Platform$file.sep), force=TRUE)
  }
  print(sprintf("clean %s", LIBSTATIC))
  unlink(LIBSTATIC, force=TRUE)
  print(sprintf("clean %s", LIBSHARED))
  unlink(LIBSHARED, force=TRUE)
  print(sprintf("clean %s", LIB))
  unlink(LIB, force=TRUE)
  print(sprintf("clean %s", TSTTARGET))
  unlink(TSTTARGET, force=TRUE)
} else if (args[1] == "help") {
  print(sprintf("%s <target>", script_name))
  print("<target>: ")
  print("                    compile cgraph")
  print("          test      test cgraph")
  print("          clean     clean all the generated files")
  print("          distclean clean all the generated files and directories")
  print("          help      commands to this program")
} else {
  print(sprintf("%s is an unsupported command", args[1]))
  print(sprintf("use \"%s help\" to know all supported commands", script_name))
}